### DISCUSSION TITLE
How Do I Get a Handle to An Asset Created in an Asset Loader?

### DISCUSSION THREAD

**Discussion by zicklag** [bevyengine/bevy#1280] [Q&A] [answered]
I have an issue where I am making a custom asset loader that will depend on a PNG image that it uses to create a `TextureAtlass`, but I can't get a handle to the `TextureAtlas` so that I can put it in my custom Asset Struct:

```rust
            // Convert that to an asset path for the texture
            let texture_path = AssetPath::new(atlas_file_path.clone(), None);

            // get the texture handle
            let texture_handle: Handle<Texture> = load_context.get_handle(texture_path.clone());

            // Create a new atlas from the texture
            let texture_atlas = TextureAtlas::from_grid(
                texture_handle,
                Vec2::new(grid_size.0 as f32, grid_size.1 as f32),
                tiles.0 as usize,
                tiles.1 as usize,
            );

            // This works,
            // Add it as a labled asset
            load_context.set_labeled_asset::<TextureAtlas>(
                "atlas",
                LoadedAsset::new(texture_atlas).with_dependency(texture_path),
            );

            // But there is no way to get a handle to my TextureAtlass in a custom asset loader?
```

Adding it as a labeled asset works kind of well, because I can load it like this:

```rust
    let character_handle: Handle<Character> = asset_server.load("player1.character.yml");
    let atlas_handle: Handle<TextureAtlas> = asset_server.load("player1.character.yml#atlas");
```

But it is inconvenient as I would prefer that there be a field already in the `Character` struct that stores the `Handle<TextureAtlas>` so that when spawning the character I don't have to manually load two handles and pass those into a bundle.

**Comment by inodentry** [comment]
You can store the handles in a resource for them to be available for you to use later from any system.

Make a resource type:

```rust
struct CharacterAssets {
    texture: Handle<Texture>,
    atlas: Handle<TextureAtlas>,
}
```

In your setup system (the one you showed, for loading the assets), add a `commands: &mut Commands` system parameter, and use it to create the resource:

```rust
commands.insert_resource(CharacterAssets { texture, atlas });
```

Then, in the system that spawns your character, you can just get the handles from the resource:

```rust
fn my_spawn_system(handles: Res<CharacterAssets>, /* ... */) {
    // use `handles.texture` and `handles.atlas`
}
```

**Reply by zicklag** [reply]
I could do that, but that still impacts the user experience because the user still has to load two assets instead of one. I'm trying to make a Bevy plugin and I wanted it to be as easy to use as possible. Still it's not that bad, I just wanted to make sure there wasn't a way to somehow produce that `Handle<TextureAtlas>` from inside the asset loader's `load` function.

**Reply by inodentry** [reply]
Oh, I understand. I've been wondering about similar things, about how the ergonomics of using assets can be improved.

Earlier today I started a discussion about this, first on discord, and now here: #1307 . Hopefully we will have better APIs for asset processing soon!

**Comment by mockersf** [comment, marked as answer]
You can look at how the [gltf loader](https://github.com/bevyengine/bevy/blob/master/crates/bevy_gltf/src/loader.rs) works.

It loads a struct that contains other handles: https://github.com/bevyengine/bevy/blob/bff44f76ec8a532b0cac88fa02cc2583a2e860e1/crates/bevy_gltf/src/lib.rs#L29-L31

for examples, for materials, it first loads all the textures with `load_context.get_handle`, and keeps the handles:
https://github.com/bevyengine/bevy/blob/bff44f76ec8a532b0cac88fa02cc2583a2e860e1/crates/bevy_gltf/src/loader.rs#L271-L305

And in the end, all handles are saved in the default object loaded for the asset:
https://github.com/bevyengine/bevy/blob/bff44f76ec8a532b0cac88fa02cc2583a2e860e1/crates/bevy_gltf/src/loader.rs#L253

**Reply by zicklag** [reply]
The issue I'm having though is that I need to get a handle to an asset that I created, not one that I can `load_context.get_handle`. In my use-case, my loader get's a texture handle using `load_context.get_handle`, but then it _creates_ a `TextureAtlas`. The issue is getting a handle to the texture atlas I created, which is impossible without having access to an `Assets<TextureAtlas>` object, which isn't present in the `load_context`.

The `load_context.get_handle` strategy works when there is a path to the asset I need a handle to, but when my loader is the source of the asset, there is no path to it. Wait! Unless... Maybe I can use `load_context.get_handle("path/to/the/asset/I/am/currently/loading#my-atlas")` where `my-atlas` is the asset that I thought I was unable to get a handle to. :thinking: So the asset get's a handle to itself, essentially, or at least one of it's own named sub-assets. I'll have to try that.

**Reply by mockersf** [reply]
oh, then on master, `set_labeled_asset` from the `load_context` now returns an `Handle` :)
https://github.com/bevyengine/bevy/blob/21794fe6df65002521823f2f20eff5ab0a158373/crates/bevy_asset/src/loader.rs#L92

**Reply by zicklag** [reply]
Perfect!

**Comment by fantasyRqg** [comment]
@zicklag Would you mind provide a sample code ?

**Reply by zicklag** [reply]
I haven't done this in a while, and I don't have any sample code lying around, unless you can find anything useful in this code [here](https://github.com/fishfolk/punchy/blob/ab48fe1545bba2c9bc8d0c971d8d7fc4efd7f74e/src/assets.rs#L52C25-L52C25) which is different than in the orignal post.

Otherwise see [`LoadContext::load()`](https://docs.rs/bevy/latest/bevy/asset/struct.LoadContext.html#method.load) which returns a handle.

**Reply by nicopap** [reply]
Note that this discussion is from _January 2021_, almost three years ago. Bevy has changed greatly since then. Take anything said in this thread with a grain of salt (actually more than a grain, you should probably ignore it).

As zicklag said, you would want `LoadContext::load`.

**Reply by fantasyRqg** [reply]
Thanks a lot. `load_context.add_labeled_asset("atlas".to_string(), atlas)` solved my problem. At first, I want to add something to asset `TextureAtlas` and I was not aware of bevy can use `type` to communicate with engine buffer (`Res<Assets<TextureAtlas>>`) out of `AssetLoader`.
